### PR TITLE
[Datasets] Add feature guide section for configuring ``batch_size`` in ``.map_batches()``

### LINKED ISSUE #29116: [AIR - Datasets] Add feature guide for configuring ``batch_size`` for ``ds.map_batches()``
Users may need guidance on configuring ``batch_size`` for `ds.map_batches()``, since the hardcoded default of 4096 will be suboptimal for wide tables and datasets containing large images, and the semantics of parallelism vs. batch size got more complicated with the block bundling PR.

### PR DESCRIPTION
**Pull request by clarkzinzow**
Users may need guidance on configuring `batch_size` for `ds.map_batches()`, since the hardcoded default of 4096 will be suboptimal for wide tables and datasets containing large images, and the semantics of parallelism vs. batch size got more complicated with the block bundling PR. This PR adds a section for configuring ``batch_size`` to the "Transforming Datasets" feature guide.

## TODOs

- [ ] (Follow-up, after batch size is configurable for preprocessors) Add code example showing parallelism vs. batch size, where batch size is tweaked?

## Related issue number

Closes #29116

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/data/transforming-datasets.rst [hunk 1]
```diff
@@ -301,6 +301,40 @@ The following output types are allowed for per-row UDFs (e.g.,
...
+
+.. note::
+  The default ``batch_size`` of ``4096`` may be too large for wide tables (many columns)
+  or datasets containing large images.
+
+Datasets will also bundle multiple blocks together for a single mapper task in order
+to better satisfy ``batch_size``, so if ``batch_size`` is a lot larger than your Dataset
+blocks (e.g. if your dataset was created with too large of a ``parallelism`` and/or the
+``batch_size`` is set to too large of a value), the number of parallel mapper tasks may
+be less than expected and you may be leaving some throughput on the table.
```
Comment (matthewdeng) [inline doc/source/data/transforming-datasets.rst:332]: Is this just referring to block coalescing? If so, won't the number of mapper tasks only be limited by `batch_size` and not the block size when parallelism is large?

```
num_tasks = num_rows/max(batch_size, block_size)
```

In other words, it's not really that the parallelism is too large, but still that the batch size is too large.

File: doc/source/data/transforming-datasets.rst [hunk 2]
```diff
@@ -301,6 +301,40 @@ The following output types are allowed for per-row UDFs (e.g.,
...
 
+----------------------
+Configuring Batch Size
+----------------------
+
+:meth:`ds.map_batches() <ray.data.Dataset.map_batches>` is the canonical parallel
+transformation API for Datasets: it launches parallel tasks over the underlying Datasets
+blocks and maps UDFs over batches of data within those tasks, allowing the UDF to
+implement vectorized operations on batches. An important parameter to
+set is the ``batch_size``, which controls the size of batches provided to the UDF.
```
Comment (matthewdeng) [inline doc/source/data/transforming-datasets.rst:312]:
```suggestion
set is ``batch_size``, which controls the size of batches provided to the UDF.
```

**Review by matthewdeng — APPROVED** [review]

File: doc/source/data/transforming-datasets.rst (hunk 1 above)
Comment (clarkzinzow) [reply doc/source/data/transforming-datasets.rst:332]: >If so, won't the number of mapper tasks only be limited by batch_size and not the block size when parallelism is large?

Hmm the parallelism determines the block size, so that's a bit of a tautology.

>In other words, it's not really that the parallelism is too large, but still that the batch size is too large.

But the default parallelism can be set to a large value (relative to the batch size) automatically by Datasets, so if the user is relying the default parallelism but is manually setting the batch size (e.g. because they're tailored the batch size to what's needed to saturate a GPU for prediction), the `parallelism` may be the knob that they need to tune.

Comment (clarkzinzow) [reply doc/source/data/transforming-datasets.rst:332]: @matthewdeng Does that make sense? Any ideas for how we can improve this section? Otherwise I'm planning on merging this.

**Comment by clarkzinzow** [comment]
@matthewdeng Ping for another review, I updated the feature guide section after @c21's PR: https://github.com/ray-project/ray/pull/29971

cc @ray-project/ray-docs for code-owner review

File: doc/source/data/transforming-datasets.rst [hunk 3]
```diff
@@ -301,6 +301,49 @@ The following output types are allowed for per-row UDFs (e.g.,
...
+
+If you specify a ``batch_size`` that's larger than your ``Dataset`` blocks, Datasets
+will bundle multiple blocks together for a single mapper task in order to better satisfy
+``batch_size``. If ``batch_size`` is a lot larger than your ``Dataset`` blocks (e.g. if
+your dataset was created with too large of a ``parallelism`` and/or the ``batch_size``
+is set to too large of a value for your dataset), the number of parallel mapper tasks
+may be less than expected and you may be leaving some throughput on the table.
+
+If your ``Dataset`` blocks are smaller than your ``batch_size`` and you want to increase
+transformation parallelism, decrease your ``batch_size`` to prevent this block bundling.
```
Comment (c21) [inline doc/source/data/transforming-datasets.rst:338]: maybe `transformation` -> `map_batches`? Just want to be specific.

File: doc/source/data/transforming-datasets.rst [hunk 4]
```diff
@@ -301,6 +301,49 @@ The following output types are allowed for per-row UDFs (e.g.,
...
+If you think that your ``Dataset`` blocks are too small, try decreasing ``parallelism``
+during the read to create larger blocks.
+
+.. note::
+  The size of the batches provided to the UDF may be smaller than the provided
+  ``batch_size`` if ``batch_size`` doesn't evenly divide the block(s) sent to a given
+  map task.
```
Comment (c21) [inline doc/source/data/transforming-datasets.rst:345]: `map task` -> `task`? To be consistent within this context?

**Review by c21 — APPROVED** [review]
LGTM with some minor comments.

File: doc/source/data/transforming-datasets.rst [hunk 5]
```diff
@@ -301,6 +301,49 @@ The following output types are allowed for per-row UDFs (e.g.,
...
+blocks and maps UDFs over batches of data within those tasks, allowing the UDF to
+implement vectorized operations on batches. An important parameter to
+set is ``batch_size``, which controls the size of the batches provided to the UDF.
+
+.. literalinclude:: ./doc_code/transforming_datasets.py
+  :language: python
+  :start-after: __configuring_batch_size_begin__
+  :end-before: __configuring_batch_size_end__
+
+Increasing ``batch_size`` can result in faster execution by better leveraging SIMD
```
Comment (jianoaix) [inline doc/source/data/transforming-datasets.rst:321]: If just for SIMD, batch_size actually needs to be small enough to fit the CPU cache right?

File: doc/source/data/transforming-datasets.rst (hunk 1 above)
Comment (matthewdeng) [reply doc/source/data/transforming-datasets.rst]: Ah reading this over I _think_ that I was referring to the last line:

> the number of parallel mapper tasks may be less than expected 

In which case lowering `parallelism` (or increasing `block_size`) won't increase the number of parallel tasks?

Comment (clarkzinzow) [reply doc/source/data/transforming-datasets.rst]: This is more about making the transformation-time parallelism respect the read-time `parallelism` argument than increasing transformation-time parallelism, so maybe this doesn't need a call out?

The user expects the number of parallel mapper tasks to be what they give for read `parallelism`, but if `parallelism` is set too high, this can result in blocks that are much smaller than `batch_size`, and since we bundle blocks to try to achieve a full `batch_size` batch in our map tasks, the map parallelism can unexpectedly be a lot smaller than the read `parallelism`, which is what we're trying to call out here. All that decreasing `parallelism` would do in this case is result make the read-time parallelism more closely match the transformation-time parallelism.

Comment (matthewdeng) [reply doc/source/data/transforming-datasets.rst]: I see, my confusion was around what the user would consider "expected"  behavior, and (maybe due to my existing knowledge) I assumed they would be limited by batch size. I guess the other thing about this paragraph is that (unlike the following paragraph) there's no concrete action for the user to take.

I don't have a concrete suggestion to improve this, so feel free to consider this comment non-blocking.

File: doc/source/data/transforming-datasets.rst (hunk 5 above)
Comment (clarkzinzow) [reply doc/source/data/transforming-datasets.rst:321]: I should probably reword this to "better leveraging vectorized operations" or something. There's the SIMD level of vectorization and there's also the native-core level of vectorization, where the latter refers to a data batch that's processed in a tight C loop, in the native-core of a particular data processing library (e.g. NumPy, Arrow, etc.).

Along with the baseline universal truth that "looping in a native language is a lot faster than looping in Python), there are also a few different levels of hardware constraints here:
1. **SIMD:** parallelized operation over 256 bits of the batch - 4 64-bit elements
2. **CPU cache line:** 64 bytes - 8 64-bit elements
3. **CPU caches:** 32 KiB L1 cache - 4096 64-bit elements, 256 KiB L2 cache - 32768 64-bit elements, 16 MiB - 2097152 64-bit elements

Assuming that the CPU does some quite aggressive cache line prefetching into the CPU caches when it detects our sequential/strided access, I would imagine that the stream of elements into the e.g. AVX2 instructions remains pretty hot.

So increasing from a batch size of 1 to 4 could lead to better SIMD fulfillment, increasing from 4 to 8 could lead to better cache line fulfillment, and increasing from 8 to e.g. 32, 64, 128 may better line up with cache line prefetching. It's difficult to be concrete without profiling a particular workload and setup, so maybe sticking with "better leveraging vectorized operations" is a good high-level umbrella statement.

File: doc/source/data/transforming-datasets.rst (hunk 1 above)
Comment (clarkzinzow) [reply doc/source/data/transforming-datasets.rst]: >I see, my confusion was around what the user would consider "expected" behavior, and (maybe due to my existing knowledge) I assumed they would be limited by batch size.

The commonly expected behavior I've heard from users is that their transformation parallelism is equal to their read parallelism, so this paragraph is mostly explaining that discrepancy for why those might not match.

>I guess the other thing about this paragraph is that (unlike the following paragraph) there's no concrete action for the user to take.

The action item of decreasing the batch size and/or decreasing the parallelism is somewhat implied by the "too large" comment, but since this is trying to alert about a possibly surprising discrepancy, not sure if we need a super explicit action item. 🤔 

>I don't have a concrete suggestion to improve this, so feel free to consider this comment non-blocking.

I'm going to leave this as-is for now so we have something in the docs to point to when talking about this parallelism discrepancy, but agreed that we might want to revisit this. I'm thinking that a lot of this feature guide will get reworked with the new execution planner + model, so we can revisit this at that time.
